File: src/main/java/com/paragon/client/systems/ui/panel/impl/module/ModuleButton.java
package com.paragon.client.systems.ui.panel.impl.module;

import com.paragon.api.util.render.GuiUtil;
import com.paragon.api.util.render.RenderUtil;
import com.paragon.api.util.render.TextRenderer;
import com.paragon.client.systems.ui.animation.Animation;
import com.paragon.client.systems.ui.panel.PanelGUI;
import com.paragon.client.systems.ui.panel.impl.Panel;
import com.paragon.client.systems.ui.panel.impl.setting.*;
import com.paragon.client.systems.module.Module;
import com.paragon.client.systems.module.impl.client.Colours;
import com.paragon.client.systems.module.impl.client.ClickGUI;
import com.paragon.client.systems.module.setting.Setting;
import net.minecraft.client.Minecraft;
import org.lwjgl.opengl.GL11;

import java.awt.*;
import java.util.ArrayList;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * @author Wolfsurge
 */
public class ModuleButton implements TextRenderer {

    // Opening / Closing animation
    private final Animation animation;

    // The parent panel
    private final Panel panel;

    // The module
    private final Module module;
    // A list of all setting components
    private final ArrayList<SettingComponent<?>> settingComponents = new ArrayList<>();
    // The offset and height
    public float offset, height;

    public ModuleButton(Panel panel, Module module, float offset, float height) {
        this.panel = panel;
        this.module = module;
        this.offset = offset;
        this.height = height;

        float settingOffset = height;

        // Add settings. Please make a PR if you want to make this look better.
        for (Setting<?> setting : getModule().getSettings()) {
            if (setting.getValue() instanceof Boolean) {
                settingComponents.add(new BooleanComponent(this, (Setting<Boolean>) setting, settingOffset, height));
                settingOffset += height;
            } else if (setting.getValue() instanceof AtomicInteger) {
                settingComponents.add(new KeybindComponent(this, (Setting<AtomicInteger>) setting, settingOffset, height));
                settingOffset += height;
            } else if (setting.getValue() instanceof Number) {
                settingComponents.add(new SliderComponent(this, (Setting<Number>) setting, settingOffset, height));
                settingOffset += height;
            } else if (setting.getValue() instanceof Enum<?>) {
                settingComponents.add(new ModeComponent(this, (Setting<Enum<?>>) setting, settingOffset, height));
                settingOffset += height;
            } else if (setting.getValue() instanceof Color) {
                settingComponents.add(new ColourComponent(this, (Setting<Color>) setting, settingOffset, height));
                settingOffset += height;
            }
        }

        animation = new Animation(100, false, () -> ClickGUI.easing.getValue());
    }

    public void renderModuleButton(int mouseX, int mouseY) {
        // Set animation speed
        animation.time = ClickGUI.animationSpeed.getValue();

        // Header
        RenderUtil.drawRect(getPanel().getX(), getOffset(), getPanel().getWidth(), getHeight(), isMouseOver(mouseX, mouseY) ? new Color(23, 23, 23).brighter().getRGB() : new Color(23, 23, 23).getRGB());

        GL11.glPushMatrix();
        // Scale it
        GL11.glScalef(0.8f, 0.8f, 0.8f);

        float scaleFactor = 1.25f;

        // Render the module's name
        renderText(getModule().getName(), (getPanel().getX() + 3) * scaleFactor, (getOffset() + 4f) * scaleFactor, getModule().isEnabled() ? Colours.mainColour.getValue().getRGB() : -1);

        // Render some dots at the side if we have more settings than just the keybind
        if (module.getSettings().size() > 1) {
            Minecraft.getMinecraft().fontRenderer.drawStringWithShadow("...", (getPanel().getX() + getPanel().getWidth() - Minecraft.getMinecraft().fontRenderer.getStringWidth("...") - 1) * scaleFactor, (getOffset() + 1.5f) * scaleFactor, -1);
        }

        GL11.glPopMatrix();

        // Refresh settings
        refreshSettingOffsets();

        if (animation.getAnimationFactor() > 0) {
            // Render settings
            settingComponents.forEach(settingComponent -> {
                if (settingComponent.getSetting().isVisible()) {
                    settingComponent.renderSetting(mouseX, mouseY);
                    RenderUtil.drawRect(getPanel().getX(), getOffset() + settingComponent.getOffset(), 1, (settingComponent instanceof ColourComponent ? getHeight() : settingComponent.getHeight()), Colours.mainColour.getValue().getRGB());
                }
            });
        }

        if (isMouseOver(mouseX, mouseY)) {
            PanelGUI.tooltip = module.getDescription();
        }
    }

    /**
     * Check if the mouse is over the module button
     *
     * @param mouseX The mouse's X
     * @param mouseY The mouse's Y
     * @return If the mouse is over the module button
     */
    public boolean isMouseOver(int mouseX, int mouseY) {
        return GuiUtil.mouseOver(getPanel().getX(), getOffset(), getPanel().getX() + getPanel().getWidth(), getOffset() + getHeight(), mouseX, mouseY);
    }

    public void mouseClicked(int mouseX, int mouseY, int mouseButton) {
        if (mouseButton == 0) {
            if (isMouseOver(mouseX, mouseY)) {
                // Toggle the module
                getModule().toggle();
            }
        } else if (mouseButton == 1) {
            if (isMouseOver(mouseX, mouseY)) {
                // Expand the settings
                animation.setState(!(animation.getAnimationFactor() > 0));
            }
        }

        if (animation.getAnimationFactor() > 0) {
            // Mouse clicked
            for (SettingComponent<?> settingComponent : settingComponents) {
                if (settingComponent.getSetting().isVisible()) {
                    settingComponent.mouseClicked(mouseX, mouseY, mouseButton);
                }
            }
        }
    }

    public void mouseReleased(int mouseX, int mouseY, int mouseButton) {
        for (SettingComponent<?> settingComponent : settingComponents) {
            if (settingComponent.getSetting().isVisible()) {
                settingComponent.mouseReleased(mouseX, mouseY, mouseButton);
            }
        }
    }

    public void keyTyped(char keyTyped, int keyCode) {
        for (SettingComponent<?> settingComponent : settingComponents) {
            if (settingComponent.getSetting().isVisible()) {
                settingComponent.keyTyped(keyTyped, keyCode);
            }
        }
    }

    public void refreshSettingOffsets() {
        float settingOffset = height;

        // EW
        for (SettingComponent<?> settingComponent : settingComponents) {
            if (settingComponent.getSetting().isVisible()) {
                settingComponent.setOffset(settingOffset);
                settingOffset += settingComponent.getHeight() * ClickGUI.animation.getValue().getAnimationFactor((float) animation.getAnimationFactor());

                if (settingComponent.animation.getAnimationFactor() > 0) {
                    float subsettingOffset = settingComponent.getOffset() + settingComponent.getHeight();
                    for (SettingComponent<?> settingComponent1 : settingComponent.getSettingComponents()) {
                        if (settingComponent1.getSetting().isVisible()) {
                            settingComponent1.setOffset(subsettingOffset);
                            subsettingOffset += settingComponent1.getHeight() * ClickGUI.animation.getValue().getAnimationFactor((float) settingComponent.animation.getAnimationFactor());
                            settingOffset += settingComponent1.getHeight() * settingComponent.animation.getAnimationFactor();
                        }
                    }
                }
            }
        }
    }

    /**
     * Gets the parent panel
     *
     * @return The parent panel
     */
    public Panel getPanel() {
        return panel;
    }

    /**
     * Gets the module
     *
     * @return The module
     */
    public Module getModule() {
        return module;
    }

    /**
     * Gets the offset
     *
     * @return The offset
     */
    public float getOffset() {
        return offset;
    }

    /**
     * Gets the height of the button
     *
     * @return The height
     */
    public float getHeight() {
        return height;
    }

    /**
     * Gets the height of the button and it's settings
     *
     * @return The absolute height
     */
    public float getAbsoluteHeight() {
        float settingHeight = 0;

        for (SettingComponent<?> settingComponent : settingComponents) {
            if (settingComponent.getSetting().isVisible()) {
                settingHeight += settingComponent.getHeight();

                if (settingComponent.animation.getAnimationFactor() > 0) {
                    for (SettingComponent<?> settingComponent1 : settingComponent.getSettingComponents()) {
                        if (settingComponent1.getSetting().isVisible()) {
                            settingHeight += settingComponent1.getHeight() * settingComponent.animation.getAnimationFactor();
                        }
                    }
                }
            }
        }

        return (float) (height + (settingHeight * animation.getAnimationFactor()));
    }

    /**
     * Gets whether the component is expanded or not
     *
     * @return Whether the component is expanded or not
     */
    public boolean isExpanded() {
        return animation.getAnimationFactor() > 0;
    }
}

File: src/main/kotlin/com/paragon/client/ui/configuration/old/impl/module/ModuleButton.kt
package com.paragon.client.ui.configuration.old.impl.module

import com.paragon.api.module.Module
import com.paragon.api.setting.Bind
import com.paragon.api.setting.Setting
import com.paragon.api.util.render.RenderUtil.drawRect
import com.paragon.api.util.render.font.FontUtil.drawStringWithShadow
import com.paragon.client.systems.module.impl.client.ClickGUI.animationSpeed
import com.paragon.client.systems.module.impl.client.ClickGUI.easing
import com.paragon.client.systems.module.impl.client.Colours
import com.paragon.client.ui.configuration.old.OldPanelGUI
import com.paragon.client.ui.configuration.old.OldPanelGUI.Companion.isInside
import com.paragon.client.ui.configuration.old.impl.Panel
import com.paragon.client.ui.configuration.old.impl.setting.*
import me.surge.animation.Animation
import net.minecraft.client.Minecraft
import org.lwjgl.opengl.GL11
import java.awt.Color
import java.util.function.Consumer

/**
 * @author Wolfsurge
 */
class ModuleButton(

    // The parent panel
    val panel: Panel,

    // The module
    val module: Module,

    // The offset and height
    var offset: Float,

    var height: Float
) {
    // Opening / Closing animation
    private val animation: Animation

    // A list of all setting components
    private val settingComponents = ArrayList<SettingComponent<*>>()

    init {
        var settingOffset = height

        // Add settings. Please make a PR if you want to make this look better.
        for (setting in module.settings) {
            if (setting.value is Boolean) {
                settingComponents.add(BooleanComponent(this, setting as Setting<Boolean>, settingOffset, height))
                settingOffset += height
            }
            else if (setting.value is Bind) {
                settingComponents.add(KeybindComponent(this, setting as Setting<Bind?>, settingOffset, height))
                settingOffset += height
            }
            else if (setting.value is Number) {
                settingComponents.add(SliderComponent(this, setting as Setting<Number?>, settingOffset, height))
                settingOffset += height
            }
            else if (setting.value is Enum<*>) {
                settingComponents.add(ModeComponent(this, setting as Setting<Enum<*>>, settingOffset, height))
                settingOffset += height
            }
            else if (setting.value is Color) {
                settingComponents.add(ColourComponent(this, setting as Setting<Color>, settingOffset, height))
                settingOffset += height
            }
        }
        animation = Animation(animationSpeed::value, false, easing::value)
    }

    fun renderModuleButton(mouseX: Int, mouseY: Int) {
        // Header
        drawRect(panel.x, offset, panel.width, height, if (isMouseOver(mouseX, mouseY)) Color(23, 23, 23).brighter().rgb else Color(23, 23, 23).rgb)
        GL11.glPushMatrix()
        // Scale it
        GL11.glScalef(0.8f, 0.8f, 0.8f)
        val scaleFactor = 1.25f

        // Render the module's name
        drawStringWithShadow(module.name, (panel.x + 3) * scaleFactor, (offset + 4f) * scaleFactor, if (module.isEnabled) Colours.mainColour.value.rgb else -1)

        // Render some dots at the side if we have more settings than just the keybind
        if (module.settings.size > 1) {
            Minecraft.getMinecraft().fontRenderer.drawStringWithShadow("...", (panel.x + panel.width - Minecraft.getMinecraft().fontRenderer.getStringWidth("...") - 1) * scaleFactor, (offset + 1.5f) * scaleFactor, -1)
        }
        GL11.glPopMatrix()

        // Refresh settings
        refreshSettingOffsets()
        if (animation.getAnimationFactor() > 0) {
            // Render settings
            settingComponents.forEach(Consumer { settingComponent: SettingComponent<*> ->
                if (settingComponent.setting.isVisible()) {
                    settingComponent.renderSetting(mouseX, mouseY)
                    drawRect(panel.x, offset + settingComponent.offset, 1f, if (settingComponent is ColourComponent) height else settingComponent.height, Colours.mainColour.value.rgb)
                }
            })
        }
        if (isMouseOver(mouseX, mouseY)) {
            OldPanelGUI.tooltip = module.description
        }
    }

    /**
     * Check if the mouse is over the module button
     *
     * @param mouseX The mouse's X
     * @param mouseY The mouse's Y
     * @return If the mouse is over the module button
     */
    fun isMouseOver(mouseX: Int, mouseY: Int): Boolean {
        return isInside(panel.x, offset, panel.x + panel.width, offset + height, mouseX, mouseY)
    }

    fun mouseClicked(mouseX: Int, mouseY: Int, mouseButton: Int) {
        if (mouseButton == 0) {
            if (isMouseOver(mouseX, mouseY)) {
                // Toggle the module
                module.toggle()
            }
        }
        else if (mouseButton == 1) {
            if (isMouseOver(mouseX, mouseY)) {
                // Expand the settings
                animation.state = animation.getAnimationFactor() <= 0
            }
        }
        if (animation.getAnimationFactor() > 0) {
            // Mouse clicked
            for (settingComponent in settingComponents) {
                if (settingComponent.setting.isVisible()) {
                    settingComponent.mouseClicked(mouseX, mouseY, mouseButton)
                }
            }
        }
    }

    fun mouseReleased(mouseX: Int, mouseY: Int, mouseButton: Int) {
        for (settingComponent in settingComponents) {
            if (settingComponent.setting.isVisible()) {
                settingComponent.mouseReleased(mouseX, mouseY, mouseButton)
            }
        }
    }

    fun keyTyped(keyTyped: Char, keyCode: Int) {
        for (settingComponent in settingComponents) {
            if (settingComponent.setting.isVisible()) {
                settingComponent.keyTyped(keyTyped, keyCode)
            }
        }
    }

    private fun refreshSettingOffsets() {
        var settingOffset = height

        // EW
        for (settingComponent in settingComponents) {
            if (settingComponent.setting.isVisible()) {
                settingComponent.offset = settingOffset
                settingOffset += (settingComponent.height * animation.getAnimationFactor()).toFloat()
                if (settingComponent.animation.getAnimationFactor() > 0) {
                    var subsettingOffset = settingComponent.offset + settingComponent.height
                    for (settingComponent1 in settingComponent.settingComponents) {
                        if (settingComponent1.setting.isVisible()) {
                            settingComponent1.offset = subsettingOffset
                            subsettingOffset += settingComponent1.height * settingComponent.animation.getAnimationFactor().toFloat()
                            settingOffset += (settingComponent1.height * settingComponent.animation.getAnimationFactor()).toFloat()
                        }
                    }
                }
            }
        }
    }

    /**
     * Gets the height of the button and it's settings
     *
     * @return The absolute height
     */
    val absoluteHeight: Float
        get() {
            var settingHeight = 0f
            for (settingComponent in settingComponents) {
                if (settingComponent.setting.isVisible()) {
                    settingHeight += settingComponent.height
                    if (settingComponent.animation.getAnimationFactor() > 0) {
                        for (settingComponent1 in settingComponent.settingComponents) {
                            if (settingComponent1.setting.isVisible()) {
                                settingHeight += (settingComponent1.height * settingComponent.animation.getAnimationFactor()).toFloat()
                            }
                        }
                    }
                }
            }
            return (height + settingHeight * animation.getAnimationFactor()).toFloat()
        }

    /**
     * Gets whether the component is expanded or not
     *
     * @return Whether the component is expanded or not
     */
    val isExpanded: Boolean
        get() = animation.getAnimationFactor() > 0
}